### PR TITLE
Add Support for longPress to MenuTrigger

### PR DESCRIPTION
**Pull request by lishichengyan**
Closes https://github.com/adobe/react-spectrum/issues/780

Add longPress support to `MenuTrigger`. 

Screenshot after running `yarn jest`:

Screen Shot 2021-12-11 at 2 51 43 AM

Acknowledgement: I'd like to thank the author @Nishan (https://github.com/intergalacticspacehighway) for his [awesome work](https://github.com/adobe/react-spectrum/pull/789/files?authenticity_token=7yQC%2FT4VLYJzb9y0wykjKx%2Bqhzy7io%2BBXL%2FWbiI1ivRILlZWg%2BBEXK7mgRuLbfdEZqCW8SsnDhTwI8apM%2F9v0Q%3D%3D&file-filters%5B%5D=.js&file-filters%5B%5D=.ts). My PR is based on his prior contribution. 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:
Open the storybook, navigate to MenuTrigger -> 'MenuTrigger with trigger = "longPress"', verify the following:
1. can open the menu via long press
2. can open the menu via space
3. can open the menu via enter
4. can open the menu via Alt+ArrowUp, if no selectedKeys specified, the last menu item is autofocused
5. can open the menu via Alt+ArrowDown, if no selectedKeys specified, the first menu item is autofocused

A screenshot:

Screen Shot 2021-12-11 at 3 05 12 AM

There are also tests in `packages/@react-spectrum/menu/test/MenuTrigger.test.js` to cover these cases.

## 🧢 Your Project:
NA

### CODE REVIEW

File: packages/@react-aria/menu/src/useMenuTrigger.ts [hunk 1]
```diff
@@ -73,23 +78,38 @@ export function useMenuTrigger(props: MenuTriggerAriaProps, state: MenuTriggerSt
     }
   };
 
+  let {longPressProps} = useLongPress({
+    isDisabled: false,
```
Comment (LFDanLu) [inline packages/@react-aria/menu/src/useMenuTrigger.ts:82]: Shouldn't this be set to just `isDisabled` instead of `isDisabled: false`? Wouldn't this mean the long press events would fire even if `isDisabled` is true?

File: packages/@react-aria/menu/src/useMenuTrigger.ts [hunk 2]
```diff
@@ -13,14 +13,18 @@
...
 import {useOverlayTrigger} from '@react-aria/overlays';
 
 interface MenuTriggerAriaProps {
   /** The type of menu that the menu trigger opens. */
   type?: 'menu' | 'listbox',
   /** Whether menu trigger is disabled. */
-  isDisabled?: boolean
+  isDisabled?: boolean,
+  /** How menu is triggered. */
+  trigger?: MenuTriggerType
```
Comment (LFDanLu) [inline packages/@react-aria/menu/src/useMenuTrigger.ts:27]: A bit surprised that `MenuTriggerAriaProps` doesn't extend MenuTriggerProps or anything... Open question for the team: should we refactor the props for MenuTrigger?  Would help sanitize `useMenuTriggerState` props in the process.

**Comment by snowystinger** [comment]
Are we supposed to be able to open with space/enter?
> can open the menu via space
> can open the menu via enter

I'm rereading the Issue description and I see
> Also add support for the Alt + arrow key behavior to open the menu instead of the normal Enter/Space/Arrow keys.

**Review by LFDanLu — CHANGES_REQUESTED** [review]
Look pretty good to me, just one question for you and the others more for the team.

Notes/questions to team:
- This will need docs
- Is alt + arrow keys only supposed to work when `trigger="longPress"`? If so, what about menu triggers in TableView? I assume those shouldn't have long press behavior but they'll need alt + arrow keys to remain intact

**Comment by lishichengyan** [comment]
> 

Hey @snowystinger thanks for taking a look. Because `useLongPress` uses `usePress` under the hood and `usePress` supports space & enter keys, it seems to me opening via space and enter is a default behavior? I can remove tests for them and revise the test Instructions as needed.

File: packages/@react-aria/menu/src/useMenuTrigger.ts (hunk 1 above)
Comment (lishichengyan) [reply packages/@react-aria/menu/src/useMenuTrigger.ts:82]: Makes sense, I'll fix this in my incoming commit. Thank you!

Comment (lishichengyan) [reply packages/@react-aria/menu/src/useMenuTrigger.ts]: @LFDanLu fixed in the newest commit

**Comment by snowystinger** [comment]
> Hey @snowystinger thanks for taking a look. Because useLongPress uses usePress under the hood and usePress supports space & enter keys, it seems to me opening via space and enter is a default behavior? I can remove tests for them and revise the test Instructions as needed.

No worries, I believe the use case for something like long press on a menu is to also support a click (short press) action to select. Think Photoshop tool groups, if you click one, it's selected, if you long press it, a menu is opened and you can select a different tool in the group/family. Because of this, Space and Enter should perform the short press action and the button should become selected.

These buttons can also be in ActionGroups I'm guessing, which is the reason that the arrow keys shouldn't open then unless used with the Alt modifier key. In ActionGroups, you can navigate between buttons in the group using arrow keys.

File: packages/@react-aria/menu/src/useMenuTrigger.ts [hunk 3]
```diff
@@ -13,14 +13,18 @@
...
 }
 
 interface MenuTriggerAria {
```
Comment (ktabors) [inline packages/@react-aria/menu/src/useMenuTrigger.ts:30]: When I drill into AriaButtonProps and it's children and look at it in the docs I don't see `onLongPressStart`, `onLongPressEnd`, or `onLongPress`.

File: packages/@react-spectrum/menu/test/MenuTrigger.test.js [hunk 4]
```diff
@@ -778,4 +779,133 @@ describe('MenuTrigger', function () {
...
+      });
+    });
+
+    it('should open the menu on Enter', function () {
+      const props = {onOpenChange, trigger: 'longPress'};
+      verifyMenuToggle(MenuTrigger, props, {}, (button, menu) => {
+        if (!menu) {
+          fireEvent.keyDown(button, {key: 'Enter'});
+          act(() => jest.advanceTimersByTime(DEFAULT_THRESHOLD));
+          fireEvent.keyUp(button, {key: 'Enter'});
```
Comment (ktabors) [inline packages/@react-spectrum/menu/test/MenuTrigger.test.js:805]: I'm reading this test as a long `enter` key press and when I test that in storybook instantly opens. I'm either missing what this test is testing or it doesn't really test a long `enter` key press.

File: packages/@react-spectrum/menu/test/MenuTrigger.test.js [hunk 5]
```diff
@@ -778,4 +779,133 @@ describe('MenuTrigger', function () {
...
+          triggerTouch(button);
+        }
+      });
+    });
+
+    it('should open the menu on Alt+ArrowUp', function () {
+      const props = {onOpenChange, trigger: 'longPress'};
+      verifyMenuToggle(MenuTrigger, props, {}, (button, menu) => {
+        if (!menu) {
+          fireEvent.keyDown(button, {key: 'ArrowUp', altKey: true});
```
Comment (ktabors) [inline packages/@react-spectrum/menu/test/MenuTrigger.test.js:829]: should this be more granular to control, verse alt, verse command/windows key?

File: packages/@react-spectrum/test-utils/src/events.ts [hunk 6]
```diff
@@ -22,6 +22,21 @@ export function triggerPress(element, opts = {}) {
...
 }
 
+// Triggers a "touch" event on an element.
+export function triggerTouch(element, opts = {}) {
+  fireEvent.pointerDown(element, {pointerType: 'touch', ...opts});
+  fireEvent.pointerUp(element, {pointerType: 'touch', ...opts});
+}
+
+// Triggers a "longPress" event on an element.
+const DEFAULT_THRESHOLD = 500;
```
Comment (ktabors) [inline packages/@react-spectrum/test-utils/src/events.ts:32]: DEFAULT_LONG_PRESS_TIME maybe? Should this be an export since the test used it?

File: packages/@react-spectrum/menu/test/MenuTrigger.test.js [hunk 7]
```diff
@@ -778,4 +779,133 @@ describe('MenuTrigger', function () {
     let menu2Item1 = within(menu2).getByText('Whiskey');
     expect(menu2Item1).toBeInTheDocument();
   });
+
+  describe('MenuTrigger trigger="longPress" open behavior', function () {
+    installPointerEvent();
+
+    const DEFAULT_THRESHOLD = 500;
+
+    it('should open the menu on longPress', function () {
```
Comment (ktabors) [inline packages/@react-spectrum/menu/test/MenuTrigger.test.js:788]: Can we add one or more tests showing the negative case of it not opening on click, only on longPress?

**Review by ktabors — COMMENTED** [review]
I'm happy that we are supporting long press, thank you!

I'd like to see a storybook example that mirrors the button experience that Rob brought up.

And I agree with Daniel that we need to talk about alt+arrow keys. I initial "specs" research didn't turn up the guidance I was looking for on this.

File: packages/@react-spectrum/menu/test/MenuTrigger.test.js (hunk 4 above)
Comment (lishichengyan) [reply packages/@react-spectrum/menu/test/MenuTrigger.test.js:805]: thanks for pointing this out! yes it doesn't really test a long enter key press. The expected behavior should be 'short press' enter to open the menu. I'll fix this in incoming commits.

File: packages/@react-spectrum/menu/test/MenuTrigger.test.js (hunk 5 above)
Comment (lishichengyan) [reply packages/@react-spectrum/menu/test/MenuTrigger.test.js:829]: Makes sense, i can add tests for windows key

File: packages/@react-spectrum/test-utils/src/events.ts (hunk 6 above)
Comment (lishichengyan) [reply packages/@react-spectrum/test-utils/src/events.ts:32]: Good idea. Will fix in new commits.

FYI: DEFAULT_THRESHOLD is taken from 'packages/@react-aria/interactions/src/useLongPress.ts' (https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/interactions/src/useLongPress.ts#:~:text=const%20DEFAULT_THRESHOLD%20%3D%20500%3B) but DEFAULT_LONG_PRESS_TIME is more clearer and exporting it makes sense 👍

File: packages/@react-spectrum/menu/test/MenuTrigger.test.js (hunk 7 above)
Comment (lishichengyan) [reply packages/@react-spectrum/menu/test/MenuTrigger.test.js:788]: Sure, I'll work on this this week

File: packages/@react-spectrum/menu/test/MenuTrigger.test.js (hunk 5 above)
Comment (lishichengyan) [reply packages/@react-spectrum/menu/test/MenuTrigger.test.js:829]: Hey @ktabors, I'm rereading the description of this issue. I'm wondering what's the expected behavior for command/windows key? Could you share more details? Thanks!

File: packages/dev/test-utils/src/events.ts [hunk 8]
```diff
@@ -22,6 +22,22 @@ export function triggerPress(element, opts = {}) {
...
+  fireEvent.pointerUp(element, {pointerType: 'touch', ...opts});
+}
+
+// Triggers a "longPress" event on an element.
+export const DEFAULT_LONG_PRESS_TIME = 500;
+
+export function triggerLongPress(element, opts = {}) {
+  fireEvent.pointerDown(element, {pointerType: 'touch', ...opts});
+  // @ts-ignore
+  act(() => jest.advanceTimersByTime(DEFAULT_LONG_PRESS_TIME));
```
Comment (snowystinger) [inline packages/dev/test-utils/src/events.ts:37]: Don't need the ts-ignore, you can just change the line to be like so
```suggestion
  act(() => {
    jest.advanceTimersByTime(DEFAULT_LONG_PRESS_TIME);
  });
```

Comment (lishichengyan) [reply packages/dev/test-utils/src/events.ts:37]: Thanks, will fix later

**Review by LFDanLu — DISMISSED** [review]
I think this is fine for now. There are some follow items that will need to be handled (adding docs for this new behavior and changing it so Enter/Arrow Down/Space don't open the menu for long press buttons as per https://github.com/adobe/react-spectrum/issues/780#issuecomment-660505333) but IMO those can be handled in a follow up PR

**Review by ktabors — COMMENTED** [review]
LGTM

**Review by ktabors — DISMISSED** [review]

**Review by ktabors — APPROVED** [review]
Tested in storybook and it looks good.

**Review by ktabors — APPROVED** [review]

**Review by snowystinger — APPROVED** [review]
